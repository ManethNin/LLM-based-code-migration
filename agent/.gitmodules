[submodule "bump"]
	path = bump
	url = git@github.com:LukvonStrom/bump.git
