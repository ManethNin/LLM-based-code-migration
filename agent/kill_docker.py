import docker
import time
import logging
from datetime import datetime, timedelta

# This is a script that kills longrunning docker containers

logging.basicConfig(level=logging.INFO, format='%(asctime)s - %(levelname)s - %(message)s')
logger = logging.getLogger(__name__)

def parse_docker_timestamp(timestamp_str):
    try:
        # Docker timestamps are in ISO 8601 format
        return datetime.fromisoformat(timestamp_str).replace(tzinfo=None)
        # return datetime.strptime(timestamp_str, "%Y-%m-%dT%H:%M:%S.%fZ")
    except ValueError:
        # If the above fails, try without microseconds
        return datetime.strptime(timestamp_str, "%Y-%m-%dT%H:%M:%SZ")

def kill_long_running_containers(max_runtime_minutes=35):
    client = docker.from_env()
    logger.info(f"Starting Docker container monitor. Max runtime set to {max_runtime_minutes} minutes.")
    
    while True:
        try:
            containers = client.containers.list()
            current_time = datetime.now()
            logger.info(f"Checking {len(containers)} running containers at {current_time}")
            
            for container in containers:
                try:
                    container_start_time_str = container.attrs['State']['StartedAt']
                    container_start_time = parse_docker_timestamp(container_start_time_str)
                    runtime = current_time - container_start_time

                    
                    logger.info(f"Container {container.name} running for {runtime}")
                    
                    if runtime > timedelta(minutes=max_runtime_minutes):
                        logger.warning(f"Killing container {container.name} (runtime: {runtime})")
                        try:
                            container.kill()
                            logger.info(f"Successfully killed container {container.name}")
                        except Exception as e:
                            logger.error(f"Failed to kill container {container.name}. Error: {str(e)}")
                    else:
                        logger.debug(f"Container {container.name} within runtime limit ({runtime})")
                except KeyError as e:
                    logger.error(f"Failed to process container {container.name}. Missing key: {str(e)}")
                except Exception as e:
                    logger.error(f"Unexpected error processing container {container.name}: {str(e)}")
            
            logger.info("Sleeping for 60 seconds before next check")
            time.sleep(60)
        except docker.errors.APIError as e:
            logger.error(f"Docker API error: {str(e)}")
            time.sleep(60)  # Wait before retrying
        except Exception as e:
            logger.critical(f"Unexpected error occurred: {str(e)}")
            time.sleep(60)  # Wait before retrying

if __name__ == "__main__":
    try:
        kill_long_running_containers()
    except KeyboardInterrupt:
        logger.info("Script terminated by user")
    except Exception as e:
        logger.critical(f"Unexpected error occurred: {str(e)}")