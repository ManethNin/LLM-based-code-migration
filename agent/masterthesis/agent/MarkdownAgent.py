from mistletoe import Document
from mistletoe.utils import traverse


class MarkdownAgent:
    def extract_codeblock(self, markdown):
        d = Document(markdown.split("\n"))

        def render_diff_code(token) -> str:
            if token.language == "diff" or not token.language:
                return token.content

        diffs = [
            render_diff_code(t.node)
            for t in traverse(d, include_source=True)
            if t.node.__class__.__name__ in ["CodeFence"]
        ]
        filtered_diffs = [diff for diff in diffs if diff is not None]

        assert len(filtered_diffs) < 2, "Only one diff is allowed in a markdown file"
        assert len(filtered_diffs) > 0, "No diff found in the markdown file"

        return filtered_diffs[0]
