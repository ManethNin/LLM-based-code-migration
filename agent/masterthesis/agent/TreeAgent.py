import os


def get_directory_tree(root, include_file_sizes=False, ignore_list=None):
    if ignore_list is None:
        ignore_list = [".git", ".DS_Store", "Thumbs.db", "node_modules"]

    def get_tree(path):
        tree = []
        try:
            for item in os.listdir(path):
                if item in ignore_list:
                    continue

                item_path = os.path.join(path, item)
                if os.path.isdir(item_path):
                    tree.append(
                        {
                            "name": item,
                            "type": "directory",
                            "contents": get_tree(item_path),
                        }
                    )
                else:
                    file_info = {"name": item, "type": "file"}
                    if include_file_sizes:
                        file_info["size"] = os.path.getsize(item_path)
                        file_info["modified_time"] = os.path.getmtime(item_path)
                    tree.append(file_info)
        except PermissionError:
            tree.append({"name": "Permission Denied", "type": "error"})
        except FileNotFoundError:
            tree.append({"name": "File Not Found", "type": "error"})
        return tree

    return get_tree(root)
