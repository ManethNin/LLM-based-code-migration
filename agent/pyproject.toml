[tool.poetry]
name = "masterthesis"
version = "0.1.0"
description = ""
readme = "README.md"


[tool.pdm]
[tool.pdm.dev-dependencies]
dev = [
    "pytest<9.0.0,>=7.0.0",
    "pre-commit<4.0.0,>=3.7.1",
    "sphinx<8.0.0,>=7.3.7",
    "prospector[with-bandit,with-mypy]<2.0.0,>=1.10.3",
    "syrupy<5.0.0,>=4.6.1",
    "ipykernel<7.0.0,>=6.29.4",
    "docker<8.0.0,>=7.1.0",
]

[tool.pdm.build]
includes = []
[build-system]
requires = ["pdm-backend", "cython"]
build-backend = "pdm.backend"

[tool.pytest.ini_options]
addopts = [
    "--import-mode=importlib",
]

[project]
requires-python = "<3.13,>=3.12"
name = "masterthesis"
version = "0.1.0"
description = ""
readme = "README.md"
dependencies = [
    "docker>=7.1.0",
    "pytest-mock>=3.14.0",
    "syrupy>=4.6.1",
    "python-dotenv>=1.0.1",
    "GitPython>=3.1.43",
    "langchain>=0.2.12",
    "opentelemetry-exporter-otlp>=1.26.0",
    "black[jupyter]>=24.8.0",
    "mypy>=1.11.1",
    "langchain-openai>=0.1.20",
    "langchain-anthropic>=0.1.22",
    "langgraph>=0.1.19",
    "openinference-instrumentation-langchain>=0.1.26",
    "opentelemetry-sdk>=1.26.0",
    "langchain-together>=0.1.5",
    "langchain-mistralai>=0.1.12",
    "tenacity>=8.5.0",
    "langchain-google-vertexai>=1.0.8",
    "anthropic[vertex]>=0.32.0",
    "pathvalidate>=3.2.0",
    "diff-match-patch>=20230430",
    "cython>=3.0.11",
    "langchain-nvidia-ai-endpoints>=0.2.1",
    "langchain-community>=0.2.12",
]

[project.optional-dependencies]
analysis = [
    "pandas>=2.2.2",
    "nbformat>=5.10.4",
    "graphviz>=0.20.3",
    "matplotlib>=3.9.1",
    "seaborn>=0.13.2",
    "plotly>=5.23.0",
    "jinja2>=3.1.4",
    "pyarrow>=17.0.0",
    "unidiff>=0.7.5",
    "ipykernel>=6.29.5",
    "ijson>=3.3.0",
    "tok>=0.1.14",
    "cython>=3.0.11",
    "tabulate>=0.9.0",
    "ipywidgets>=8.1.3",
    "matplotlib-venn>=1.1.1",
    "upsetplot>=0.9.0",
]
[tool.black]
extend-exclude = '''/dataset'''

[tool.pyright]
exclude = ["dataset/**/**", "dataset/*"]