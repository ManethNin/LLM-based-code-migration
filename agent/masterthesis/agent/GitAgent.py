import shutil
from pathlib import Path
import tempfile
from typing import List

import git
from unidiff import PatchSet, PatchedFile

from masterthesis.agent.DiffAgent import Patch, PatchError


class GitAgent:
    def __init__(self, repo_path: Path, commit_hash: str, github_slug: str) -> None:
        self.repo_path = repo_path
        self.commit_hash = commit_hash
        if not repo_path.exists():
            self.repo = self.raw_checkout(repo_path, github_slug)
        else:
            try:
                self.repo = git.Repo(repo_path)
            except Exception as e:
                shutil.rmtree(repo_path)
                self.repo = self.raw_checkout(repo_path, github_slug)

    def is_dirty(self) -> bool:
        return self.repo.is_dirty(untracked_files=True)

    def raw_checkout(self, repo_path: Path, github_slug: str) -> None:
        # Clone the repository
        repo = git.Repo.clone_from(f"https://github.com/{github_slug}.git", repo_path)
        repo.git.fetch("origin", self.commit_hash, depth=2)
        repo.git.checkout(self.commit_hash)
        return repo

    def get_full_diff(self) -> str:
        """
        Get the full diff of the current state of the repository compared to the last commit.

        Returns:
            str: A string containing the full diff.
        """
        if not self.is_dirty():
            return "No changes detected."

        # Get the diff for tracked files
        diff = self.repo.git.diff("HEAD", ignore_all_space=True)

        # Get the list of untracked files
        untracked_files = self.repo.untracked_files

        if untracked_files:
            diff += "\n\n{Untracked files}:\n"
            for file in untracked_files:
                diff += f"{file}\n"

        return diff

    def apply_diff(self, diff: str) -> List[str]:
        """
        Apply the full diff and return the paths of the modified files.

        Args:
            diff (str): The diff string produced by get_full_diff method.

        Returns:
            List[str]: A list of paths for the modified files.
        """
        if diff == "No changes detected.":
            return []

        modified_files = []

        # Split the diff into sections for tracked and untracked files
        tracked_diff, untracked_diff = (
            diff.split("{Untracked files}:")
            if "{Untracked files}:" in diff
            else (diff, "")
        )

        # Apply changes to tracked files

        with tempfile.NamedTemporaryFile(
            mode="w", suffix=".diff", encoding="utf-8", delete=False
        ) as f:
            f.write(tracked_diff)
            f.flush()

            patch = Patch(f.name)
            patch.apply(
                target_dir=self.repo_path.as_posix(),
                ignore_whitespace=True,
                forgiving=True,
            )

            # print(['--ignore-whitespace', '--allow-empty', f.name])
            # self.repo.git.apply(['--ignore-whitespace', '--verbose', '-3', f.name])

            # print(self.repo.is_dirty())
            # print(self.repo.untracked_files)

        # Extract modified file paths from the tracked diff

        patch = PatchSet(tracked_diff)
        modified_set = set()

        def get_file_path(patched_files: List[PatchedFile]):
            for patched_file in patched_files:
                if patched_file.target_file:
                    modified_set.add(Path(*Path(patched_file.target_file).parts[1:]))

        get_file_path(patch.modified_files)
        get_file_path(patch.added_files)
        get_file_path(patch.removed_files)

        modified_files = list(modified_set)

        # Handle untracked files
        if untracked_diff:
            for file_path in untracked_diff.strip().split("\n"):
                file_path = file_path.strip()
                if file_path:
                    full_path = self.repo_path / file_path
                    if not full_path.exists():
                        full_path.touch()
                    modified_files.append(file_path)

        return modified_files

    def discard_changes(self):
        try:
            if self.is_dirty():
                print("Repository is dirty. Discarding changes...")

                # Reset tracked files
                self.repo.head.reset(index=True, working_tree=True)

                # Clean untracked files and directories
                self.repo.git.clean("-fd")

                print("All changes have been discarded.")
        except Exception as e:
            print(f"An error occurred while discarding changes: {e}")

        if self.repo.head.commit.hexsha != self.commit_hash:
            shutil.rmtree(self.repo.working_dir)
            self.repo = self.raw_checkout(self.repo_path, self.commit_hash)
