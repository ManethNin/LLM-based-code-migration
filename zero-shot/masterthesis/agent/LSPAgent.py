import json
import logging
import os
import shutil
import signal
import tempfile
from contextlib import contextmanager
from pathlib import Path
from typing import Any, Dict, Optional, Tuple

from masterthesis.agent.aider.AdvancedDiffAgent import UnifiedDiffCoder
from masterthesis.agent.DockerAgent import DockerAgent


def extract_error_lines(docker_output: str) -> list[str]:
    error_lines = []
    RECORDING_ERRORS = False

    for line in docker_output.split("\n"):

        # if "T E S T S" in line:
        #     RECORDING_ERRORS = True
        if "BUILD FAILURE" in line:
            RECORDING_ERRORS = True

        if line.startswith("[ERROR]") and RECORDING_ERRORS:
            RECORDING_ERRORS = True
        if (
            "To see the full stack trace of the errors, re-run Maven with the -e switch."
            in line
        ):
            RECORDING_ERRORS = False
            break

        if (
            "Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin"
            in line
        ):
            error_lines.append(line.strip())
            RECORDING_ERRORS = False
            break

        if line.strip() == "[ERROR]":
            # line is empty
            continue

        if RECORDING_ERRORS and "[ERROR]" in line:
            error_lines.append(line.strip())

    if "[INFO] BUILD SUCCESS" in docker_output:
        print("BUILD SUCCESS")
        return []
    return error_lines


class LSPAgent:
    def __init__(self, project_path: Path) -> None:
        """
        Initializes the LSPAgent with the provided project path.

        :param project_path: Path to the project directory.
        """
        self.dockerAgent = DockerAgent(
            "ghcr.io/lukvonstrom/multilspy-java-docker:latest", project_path
        )

        # self.dockerAgent = DockerAgent(
        #     "multilspy-java-docker:latest", project_path
        # )
        self.project_path = project_path

        self.container = None

        self.repo_dir = project_path
        self.results_dir = tempfile.mkdtemp()
        self.input_dir = tempfile.mkdtemp()

    @contextmanager
    def start_container(self):
        # self.dockerAgent.pull_image()
        try:
            container, setup_stdout, setup_stderr = (
                self.dockerAgent.execute_command_with_mounts(
                    mounts={
                        self.repo_dir: {"bind": "/mnt/repo", "mode": "rw"},
                        self.results_dir: {
                            "bind": "/mnt/data",
                            "mode": "rw",
                        },
                        self.input_dir: {"bind": "/mnt/input", "mode": "rw"},
                    },
                    setup_command="/bin/bash -c 'source .env/bin/activate && python setup.py'",
                )
            )
            # Todo: validate setup_stdout and setup_stderr
            self.container = container
            yield self.container
        finally:
            if self.container is not None:
                self.dockerAgent.clean_up(self.container)
            else:
                self.dockerAgent.clean_up()

            if os.path.exists(self.results_dir):
                shutil.rmtree(self.results_dir)
            if os.path.exists(self.input_dir):
                shutil.rmtree(self.input_dir)
            # print("Skipped cleanup!")

    def validate_changes(
        self, original_file_path: Path, diff_texts: list[str]
    ) -> tuple[dict[Any, Any], dict[Any, Any]]:
        """
        Validates changes in the specified file based on the provided diff text.

        :param original_file_path: Path to the original file.
        :param diff_text: Text representing the differences.
        :return: Tuple of dictionaries with initial and edit LSP results.
        """
        return self._validate_lsp(original_file_path, diff_texts)

    def validate_file(self, original_file_path: Path) -> Dict:
        """
        Validates the specified file without any changes.

        :param original_file_path: Path to the original file.
        :return: Dictionary with initial LSP results.
        """
        return self._validate_lsp(original_file_path)

    def prepare_diffs(self, diffs: list[str]):
        coder = UnifiedDiffCoder(repo_dir=self.repo_dir)
        for diff_text in diffs:
            paths = coder.get_paths(diff_text)
            for path in paths:

                output_location = Path(self.input_dir) / os.path.basename(path)
                is_valid_diff, patched_file = coder.apply_edits(diff_text)
                assert is_valid_diff, f"Diff for {path} is not valid"

                with open(output_location, "w", encoding="utf-8") as out_file_wrapper:
                    out_file_wrapper.write(patched_file)

    def _validate_lsp(
        self, original_file_path: Path, diffs: Optional[list[str]] = None
    ) -> Tuple[Dict, Optional[Dict]]:
        """
        Internal method to validate a file, optionally with changes.

        :param original_file_path: Path to the original file.
        :param diff_text: Optional text representing the differences.
        :return: Tuple of dictionaries with initial and optionally edit LSP results.
        """
        # `docker run -v "$(pwd)/lithium:/mnt/repo" -v "$(pwd)/mnt/data:/mnt/data" -v "$(pwd)/edit:/mnt/input" multilspy-java-docker -e yes -f src/main/java/com/wire/lithium/models/NewBotResponseModel.java`

        assert self.container is not None, "Container is not initialized"
        change_mode = len(diffs) > 0
        logging.debug(
            "Validating file %s in %s mode",
            original_file_path,
            "change" if change_mode else "normal",
        )

        try:

            if change_mode:
                # code_file = self.project_path / original_file_path
                self.prepare_diffs(diffs)

            parameter_text = "-e yes -f " if change_mode else "-f "

            docker_output = self.dockerAgent.execute_main_command(
                self.container,
                f"/bin/bash -c '/app/init.sh {parameter_text}{original_file_path.as_posix()}'",
            )
            # logging.info("docker_output %s", docker_output)

            if "Language Server timeout" in docker_output:
                self.dockerAgent.clean_up()
                raise Exception("Language Server timeout")

            with open(
                Path(self.results_dir) / "initial.json", "r", encoding="utf-8"
            ) as initial_file:
                initial_lsp_results = json.loads(initial_file.read())

            if change_mode:
                with open(
                    Path(self.results_dir) / "edit.json", "r", encoding="utf-8"
                ) as edit_file:
                    edit_lsp_results = json.loads(edit_file.read())
                    self.dockerAgent.clean_up()
                    return initial_lsp_results, edit_lsp_results
        finally:
            self.dockerAgent.clean_up()
        return initial_lsp_results
