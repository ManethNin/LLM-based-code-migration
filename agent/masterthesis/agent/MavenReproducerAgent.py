import logging
import os
import shutil
import tempfile
from contextlib import contextmanager
from pathlib import Path
from typing import Any, Dict, Tuple

from masterthesis.agent.aider.AdvancedDiffAgent import UnifiedDiffCoder
from masterthesis.agent.DockerAgent import DockerAgent
from masterthesis.agent.LSPAgent import extract_error_lines


class MavenReproducerAgent:
    def __init__(self, project_path: Path) -> None:

        self.dockerAgent = DockerAgent("maven:3.9.8-amazoncorretto-17", project_path)

        self.project_path = project_path

        self.container = None

        self.force_upgrade_compiler_version = False
        self.compiler_upgrade_attempted = False

        self.repo_dir = project_path
        self.results_dir = tempfile.mkdtemp()
        self.input_dir = tempfile.mkdtemp()

    @contextmanager
    def start_container(self):
        # self.dockerAgent.pull_image()
        try:
            container, setup_stdout, setup_stderr = (
                self.dockerAgent.execute_command_with_mounts(
                    mounts={
                        self.repo_dir: {"bind": "/mnt/repo", "mode": "rw"},
                        self.results_dir: {
                            "bind": "/mnt/data",
                            "mode": "rw",
                        },
                        self.input_dir: {"bind": "/mnt/input", "mode": "rw"},
                    },
                    setup_command="mkdir -p /app",
                )
            )
            print("repo at", self.repo_dir)
            # Todo: validate setup_stdout and setup_stderr
            self.container = container
            yield self.container
        finally:
            if self.container is not None:
                self.dockerAgent.clean_up(self.container)
            else:
                self.dockerAgent.clean_up()

            if os.path.exists(self.results_dir):
                shutil.rmtree(self.results_dir)
            if os.path.exists(self.input_dir):
                shutil.rmtree(self.input_dir)
            # print("Skipped cleanup!")

    def compile_maven_with_full_file_replace(
        self,
        file_content: str,
        file_path: str,
        run_tests: bool = True,
        timeout: int = 1800,
    ) -> Tuple[Tuple[bool, bool], str, dict]:
        with open(file_path, "w", encoding="utf-8") as out_file_wrapper:
            out_file_wrapper.write(file_content)

        (compile, test), error_text = self._compile_maven(run_tests, timeout)

        return (compile, test), error_text, {file_path: file_content}

    def compile_maven(
        self, diffs: list[str], run_tests: bool, timeout: int = 1800
    ) -> Tuple[Tuple[bool, bool], str, dict]:

        assert self.container is not None, "Container is not initialized"

        assert isinstance(diffs, list), "Diffs are not a list"

        updated_files = {}
        try:
            coder = UnifiedDiffCoder(repo_dir=self.repo_dir)
            for diff_text in diffs:
                is_valid_diff, patched_files = coder.apply_edits(diff_text)

                assert is_valid_diff, f"Diff is not valid, {patched_files}"

                for path, patched_file in patched_files.items():

                    output_location = Path(self.repo_dir) / path

                    updated_files[path] = patched_file

                    with open(
                        output_location, "w", encoding="utf-8"
                    ) as out_file_wrapper:
                        out_file_wrapper.write(patched_file)
        except Exception as e:
            return (False, False), f"Failed to prepare diffs: {e}", {}

        (compile, test), error_text = self._compile_maven(run_tests, timeout)

        return (compile, test), error_text, updated_files

    def _compile_maven(
        self, run_tests: bool, timeout: int = 1800
    ) -> Tuple[Tuple[bool, bool], str]:
        try:
            reproduction_command = "mvn clean test -Dsurefire.printSummary=true -Dsurefire.redirectTestOutputToFile=false"

            if self.force_upgrade_compiler_version:
                reproduction_command += " -Dmaven.compiler.source=8 -Dmaven.compiler.target=8 -Dmaven.compiler.release=8 -Dmaven.compiler.testSource=8 -Dmaven.compiler.testTarget=8"

            reproduction_command += " -B"

            if not run_tests:
                reproduction_command = "mvn clean compile -DskipTests -B"

            timeout_command = f"timeout -k 10s {timeout}s {reproduction_command}"

            print(f"Running maven command {timeout_command}")
            output_code, docker_output = self.dockerAgent.execute_command(
                self.container, timeout_command, "/mnt/repo"
            )
            print("maven output_code", output_code)
            # logging.info("docker_output %s", docker_output)
            # print("docker_output", output_code, docker_output)

            if (
                "Source option 6 is no longer supported. Use 7 or later."
                in docker_output
                and not self.compiler_upgrade_attempted
            ):
                self.force_upgrade_compiler_version = True
                self.compiler_upgrade_attempted = True
                return self._compile_maven(run_tests, timeout)

            self.compiler_upgrade_attempted = False

            if (
                int(output_code) == 124
                or int(output_code) == 137
                or "timeout: sending signal TERM to command" in docker_output
            ):
                return (
                    False,
                    False,
                ), f"Maven timed out after {timeout} seconds see: {docker_output}"

            if not run_tests:
                error_lines = extract_error_lines(docker_output)

                print(
                    "has_failed output code is",
                    output_code,
                    "error line length was",
                    len(error_lines),
                )
                has_succeeded = int(output_code) == 0 and len(error_lines) < 1
                print("has_succeeded", has_succeeded, error_lines)

                if len(error_lines) > 0:
                    error_text = "\n".join(error_lines)
                    return (has_succeeded, has_succeeded), error_text.replace(
                        "/mnt/repo/", ""
                    )
                else:
                    return (has_succeeded, has_succeeded), docker_output.replace(
                        "/mnt/repo/", ""
                    )
            else:
                # Better isolator for test results
                lines = docker_output.split("\n")
                test_index = -1
                compilation_index = -1
                compilation_failed = False
                for i, line in enumerate(lines):
                    if "[INFO] Results:" in line:
                        test_index = i
                        break
                for i, line in enumerate(lines):
                    if "COMPILATION ERROR" in line:
                        compilation_failed = True
                        compilation_index = i
                        break

                if compilation_failed:
                    test_output = "\n".join(lines[(compilation_index - 1) :])
                    return (False, False), test_output

                if test_index == -1:
                    lines_without_downloads = [
                        line
                        for line in lines
                        if "Downloading" not in line and "Downloaded" not in line
                    ]
                    return (True, int(output_code) == 0), "\n".join(
                        lines_without_downloads
                    )

                test_output = "\n".join(lines[(test_index - 1) :])
                succeeded_both = int(output_code) == 0
                return (True, succeeded_both), test_output
        except Exception as e:
            return False, f"An error occurred: {str(e)}"
