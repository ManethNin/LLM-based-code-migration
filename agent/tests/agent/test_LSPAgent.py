import os
import pytest
from pathlib import Path
from masterthesis.agent.LSPAgent import LSPAgent
import tempfile
import subprocess


@pytest.fixture
def repo_dir():
    with tempfile.TemporaryDirectory() as temp_dir:
        project_path = Path(temp_dir) / "my_project"
        yield project_path


@pytest.fixture
def lsp_agent(repo_dir):
    commit_hash = "10d7545c5771b03dd9f6122bd5973a759eb2cd03"
    _ = subprocess.run(
        ["git", "clone", "https://github.com/wireapp/lithium.git", repo_dir],
        capture_output=True,
    )
    _ = subprocess.run(
        ["git", "fetch", "--depth", "2", "origin", commit_hash],
        cwd=repo_dir,
        capture_output=True,
    )
    _ = subprocess.run(
        ["git", "checkout", commit_hash], cwd=repo_dir, capture_output=True
    )
    return LSPAgent(repo_dir)


def test_validate_changes(lsp_agent, repo_dir, snapshot):
    with lsp_agent.start_container() as container:
        original_file_path = (
            "src/main/java/com/wire/lithium/models/NewBotResponseModel.java"
        )
        diff_path = Path(__file__).parent / "diff.txt"
        diff_text = diff_path.read_text()
        print(os.path.exists(repo_dir))
        result1, result2 = lsp_agent.validate_changes(original_file_path, diff_text)
        assert isinstance(result1, dict)
        assert isinstance(result2, dict)

        assert result1 == snapshot
        assert result2 == snapshot


# def test_validate_file(lsp_agent):
#   original_file_path = "/path/to/original_file.txt"
#   result = lsp_agent.validate_file(original_file_path)
#   assert isinstance(result, dict)
#   # Add more assertions based on the expected behavior of validate_file()

# def test__validate_lsp(lsp_agent):
#   original_file_path = "/path/to/original_file.txt"
#   diff_text = "some diff text"
#   result1, result2 = lsp_agent._validate_lsp(original_file_path, diff_text)
#   assert isinstance(result1, dict)
#   assert isinstance(result2, dict)
#   # Add more assertions based on the expected behavior of _validate_lsp()
