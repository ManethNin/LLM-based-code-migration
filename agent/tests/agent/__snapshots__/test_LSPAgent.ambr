# serializer version: 1
# name: test_validate_changes
  dict({
    'diagnostics': list([
      dict({
        'code': '0',
        'message': 'The project was not built since its build path is incomplete. Cannot find the class file for io.dropwizard.core.setup.Environment. Fix the build path then try building this project',
        'range': dict({
          'end': dict({
            'character': 0,
            'line': 0,
          }),
          'start': dict({
            'character': 0,
            'line': 0,
          }),
        }),
        'severity': 1,
        'source': 'Java',
      }),
      dict({
        'code': '16777563',
        'message': 'The type io.dropwizard.core.setup.Environment cannot be resolved. It is indirectly referenced from required type io.dropwizard.client.JerseyClientBuilder',
        'range': dict({
          'end': dict({
            'character': 43,
            'line': 159,
          }),
          'start': dict({
            'character': 15,
            'line': 159,
          }),
        }),
        'severity': 1,
        'source': 'Java',
      }),
    ]),
    'error': None,
    'warnings': list([
    ]),
  })
# ---
# name: test_validate_changes.1
  dict({
    'diagnostics': list([
      dict({
        'code': '0',
        'message': 'The project was not built since its build path is incomplete. Cannot find the class file for io.dropwizard.core.setup.Environment. Fix the build path then try building this project',
        'range': dict({
          'end': dict({
            'character': 0,
            'line': 0,
          }),
          'start': dict({
            'character': 0,
            'line': 0,
          }),
        }),
        'severity': 1,
        'source': 'Java',
      }),
      dict({
        'code': '16777563',
        'message': 'The type io.dropwizard.core.setup.Environment cannot be resolved. It is indirectly referenced from required type io.dropwizard.client.JerseyClientBuilder',
        'range': dict({
          'end': dict({
            'character': 43,
            'line': 159,
          }),
          'start': dict({
            'character': 15,
            'line': 159,
          }),
        }),
        'severity': 1,
        'source': 'Java',
      }),
      dict({
        'code': '33554515',
        'message': 'assets cannot be resolved to a variable',
        'range': dict({
          'end': dict({
            'character': 18,
            'line': 32,
          }),
          'start': dict({
            'character': 12,
            'line': 32,
          }),
        }),
        'severity': 1,
        'source': 'Java',
      }),
      dict({
        'code': '33554515',
        'message': 'assets cannot be resolved to a variable',
        'range': dict({
          'end': dict({
            'character': 18,
            'line': 33,
          }),
          'start': dict({
            'character': 12,
            'line': 33,
          }),
        }),
        'severity': 1,
        'source': 'Java',
      }),
      dict({
        'code': '570425394',
        'data': list([
          'assets',
        ]),
        'message': 'assets cannot be resolved',
        'range': dict({
          'end': dict({
            'character': 14,
            'line': 39,
          }),
          'start': dict({
            'character': 8,
            'line': 39,
          }),
        }),
        'severity': 1,
        'source': 'Java',
      }),
      dict({
        'code': '0',
        'message': 'The project was not built since its build path is incomplete. Cannot find the class file for io.dropwizard.core.setup.Environment. Fix the build path then try building this project',
        'range': dict({
          'end': dict({
            'character': 0,
            'line': 0,
          }),
          'start': dict({
            'character': 0,
            'line': 0,
          }),
        }),
        'severity': 1,
        'source': 'Java',
      }),
    ]),
    'error': None,
    'warnings': list([
    ]),
  })
# ---
